### PR TITLE
Add GNOI File.Stat API

### PR DESCRIPTION
**Pull request by isabelmsft**
#### Why I did it
Add gNOI API support to read file stat.
#### How I did it
Implement gNOI API to add File.Stat
Support reading file stat sonic-host-services, added in this PR: https://github.com/sonic-net/sonic-host-services/pull/142
#### How to verify it
unit testing done, E2E testing to be added
#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

File: gnmi_server/server.go [hunk 1]
```diff
@@ -158,6 +171,11 @@ func NewServer(config *Config, opts []grpc.ServerOption) (*Server, error) {
 		ReqFromMaster: ReqFromMasterDisabledMA,
 		masterEID:     uint128{High: 0, Low: 0},
 	}
+
+    // Create instances of GNMIServer and GNOIServer
+    srv.gnmiServer = &GNMIServer{Server: srv}
+    srv.gnoiServer = &GNOIServer{Server: srv}
```
Comment (ganglyu) [inline gnmi_server/server.go:177]: Use tab?

File: gnmi_server/server.go [hunk 2]
```diff
@@ -166,15 +184,16 @@ func NewServer(config *Config, opts []grpc.ServerOption) (*Server, error) {
...
 		return nil, fmt.Errorf("failed to open listener port %d: %v", srv.config.Port, err)
 	}
-	gnmipb.RegisterGNMIServer(srv.s, srv)
-	spb_jwt_gnoi.RegisterSonicJwtServiceServer(srv.s, srv)
+	gnmipb.RegisterGNMIServer(srv.s, srv.gnmiServer)
+	spb_jwt_gnoi.RegisterSonicJwtServiceServer(srv.s, srv.gnoiServer)
 	if srv.config.EnableTranslibWrite || srv.config.EnableNativeWrite {
-		gnoi_system_pb.RegisterSystemServer(srv.s, srv)
+		gnoi_system_pb.RegisterSystemServer(srv.s, srv.gnoiServer)
+		gnoi_file_pb.RegisterFileServer(srv.s, srv.gnoiServer)
```
Comment (ganglyu) [inline gnmi_server/server.go:191]: Please create systemserver and fileserver to replace this gnoi server

File: gnmi_server/server.go (hunk 1 above)
Comment (isabelmsft) [reply gnmi_server/server.go]: Fixed spacing

File: gnmi_server/server.go (hunk 2 above)
Comment (isabelmsft) [reply gnmi_server/server.go]: Thanks, changed to systemServer and fileServer

**Review by ganglyu — DISMISSED** [review]

**Comment by vvolam** [comment]
Only minor comments, otherwise lgtm. Thanks!

File: gnoi_client/gnoi_client.go [hunk 3]
```diff
@@ -62,6 +63,14 @@ func main() {
 		default:
 			panic("Invalid RPC Name")
 		}
+	case "File":
+		fc := gnoi_file_pb.NewFileClient(conn)
+		switch *rpc {
+		case "Stat":
+			fileStat(fc, ctx)
+		default:
+			panic("Invalid RPC Name")
```
Comment (zbud-msft) [inline gnoi_client/gnoi_client.go:72]: I don't think we should panic here.

File: gnoi_client/gnoi_client.go [hunk 4]
```diff
@@ -125,6 +134,25 @@ func killProcess(sc gnoi_system_pb.SystemClient, ctx context.Context) {
 	}
 }
 
+func fileStat(fc gnoi_file_pb.FileClient, ctx context.Context) {
```
Comment (zbud-msft) [inline gnoi_client/gnoi_client.go:137]: This is a comment that should be applied to this entire file. 

It seems like we throw panic a lot in this file if gnoi operation fails or during a pre check.

Is that really the intended behavior, as panic will terminate the entire process, unexpectedly shutting down telemetry/gnmi?

If we do decide that panic is the right call, we should wrap them in recover to handle them gracefully as to not cause a crash.

**Review by zbud-msft — CHANGES_REQUESTED** [review]

File: gnoi_client/gnoi_client.go (hunk 4 above)
Comment (zbud-msft) [reply gnoi_client/gnoi_client.go:137]: Resolving as these panics will not affect gnmi/telemetry server

File: gnoi_client/gnoi_client.go (hunk 3 above)
Comment (isabelmsft) [reply gnoi_client/gnoi_client.go:72]: Thanks for the suggestion, I opened an ADO 29031150 to refactor gNOI client code as suggested. I will make these changes in a future PR to compartmentalize the change

**Review by zbud-msft — DISMISSED** [review]

File: test/utils.py [hunk 5]
```diff
@@ -316,3 +316,12 @@ def gnoi_refresh_with_jwt(token):
     cmd += '-module Sonic -rpc refresh '
     ret, msg = run_cmd(cmd)
     return ret, msg
+
+def gnoi_file_stat(json_data):
+    path = os.getcwd()
+    cmd = path + '/build/bin/gnoi_client '
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply test/utils.py]: This path is passed to the `run_cmd()` func and just used for UT. No longer used in this PR, as I moved UT to golang files

**Review by vvolam — APPROVED** [review]
LGTM, Thanks!

**Pull request merged**